### PR TITLE
ARROW-8296: [C++][Dataset] Add IpcFileWriteOptions

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

**Review by pitrou — COMMENTED** [review]
LGTM on the C++ part

**Comment by bkietz** [comment]
@kou not sure how to idiomatically propagate this change to GLib; it seems there's no way to raise an exception when setting a property?

**Comment by kou** [comment]
I'll take a look it and push a fix.

**Comment by kou** [comment]
Done.

+1 on the GLib part.

**Comment by bkietz** [comment]
Thanks @kou!

File: cpp/src/arrow/dataset/file_ipc.cc [hunk 1]
```diff
@@ -185,7 +185,13 @@ Result<std::shared_ptr<FileWriter>> IpcFileFormat::MakeWriter(
 
   auto ipc_options = checked_pointer_cast<IpcFileWriteOptions>(options);
 
-  ARROW_ASSIGN_OR_RAISE(auto writer, ipc::MakeFileWriter(destination, schema));
+  // override use_threads to avoid nested parallelism
+  ipc_options->options->use_threads = false;
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/dataset/file_ipc.cc:189]: Should this in theory only be done if we have multiple files / fragments? (like the issue we had for parquet)

**Review by jorisvandenbossche — COMMENTED** [review]
Should this also be exposed in python's dataset `IpcFileWriteOptions` ? 
(can also look into pushing that)

File: cpp/src/arrow/dataset/file_ipc.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_ipc.cc:189]: That sounds acceptable but I'd like to defer it to a follow up

File: r/configure [hunk 2]
```diff
@@ -201,7 +201,10 @@ else
 fi
 
 # Write to Makevars
-sed -e "s|@cflags@|$PKG_CFLAGS|" -e "s|@libs@|$PKG_LIBS|" src/Makevars.in > src/Makevars
+sed -e "s|@cflags@|$PKG_CFLAGS|" \
```
Comment (nealrichardson) [inline r/configure:204]: Please revert this change and the one in Makevars.in. This will provoke the wrath of CRAN (and it should cause at least one nightly build failure).

File: r/src/arrow_cpp11.h [hunk 3]
```diff
@@ -157,19 +157,42 @@ struct ns {
   static SEXP arrow;
 };
 
+// Specialize this struct to define a default value to be used when NULL is given.
```
Comment (nealrichardson) [inline r/src/arrow_cpp11.h:160]: Why is this change necessary?

File: r/src/arrow_metadata.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "./arrow_types.h"
+
+#if defined(ARROW_R_WITH_ARROW)
+#include <arrow/util/key_value_metadata.h>
+
+inline std::shared_ptr<arrow::KeyValueMetadata> KeyValueMetadata__Make(
```
Comment (nealrichardson) [inline r/src/arrow_metadata.h:23]: Two thoughts:

1. We already have KeyValueMetadata support in R, so if these helpers are necessary, I'd expect to see them used in those other places too.
2. It looks like these functions are opposites, one takes a named character vector and turns it into arrow::KVM, and the other takes KVM and makes a named character vector. Can we make the function names more obvious about that?

File: r/src/recordbatchreader.cpp [hunk 5]
```diff
@@ -74,6 +76,12 @@ int ipc___RecordBatchFileReader__num_record_batches(
   return reader->num_record_batches();
 }
 
+// [[arrow::export]]
+cpp11::strings ipc___RecordBatchFileReader__metadata(
```
Comment (nealrichardson) [inline r/src/recordbatchreader.cpp:80]: It's cool that you added this but it's probably not necessary; see below for a simpler test.

File: r/tests/testthat/test-dataset.R [hunk 6]
```diff
@@ -943,10 +943,42 @@ test_that("Dataset writing: from RecordBatch", {
...
+  codec <- NULL
+  if (codec_is_available("zstd")) {
+    codec <- Codec$create("zstd")
+  }
+
+  write_dataset(ds, dst_dir, format = "feather", metadata = metadata, codec = codec)
+  expect_true(dir.exists(dst_dir))
+
+  file <- ds$filesystem$OpenInputStream(paste(dst_dir, dir(dst_dir)[[1]], sep = "/"))
+  expect_equivalent(metadata, RecordBatchFileReader$create(file)$metadata)
```
Comment (nealrichardson) [inline r/tests/testthat/test-dataset.R:961]:
```suggestion
  file <- dir(dst_dir, full.names = TRUE)[[1]]
  expect_equivalent(read_feather(file, as_data_frame = FALSE)$metadata, metadata)
```

File: r/tests/testthat/test-dataset.R [hunk 7]
```diff
@@ -943,10 +943,42 @@ test_that("Dataset writing: from RecordBatch", {
...
+      filter(integer > 6) %>%
+      summarize(mean = mean(integer))
+  )
+})
+
 test_that("Writing a dataset: Parquet format options", {
   skip_on_os("windows") # https://issues.apache.org/jira/browse/ARROW-9651
   ds <- open_dataset(csv_dir, partitioning = "part", format = "csv")
   dst_dir <- make_temp_dir()
+  dst_dir_no_truncated_timestamps <- make_temp_dir()
```
Comment (nealrichardson) [inline r/tests/testthat/test-dataset.R:981]: Why this change? I wouldn't expect this PR to affect parquet writing options, should it?

File: r/R/dataset-write.R [hunk 8]
```diff
@@ -44,7 +44,8 @@
 #' @param filesystem A [FileSystem] where the dataset should be written if it is a
 #' string file path; default is the local file system
 #' @param ... additional format-specific arguments. For available Parquet
-#' options, see [write_parquet()].
+#' options, see [write_parquet()]. For available Feather options, see
+#' [RecordBatchFileWriter$create()].
```
Comment (nealrichardson) [inline r/R/dataset-write.R:48]: I don't think that doc link will work, and if you were to do `make doc`, this will probably fail check due to a bad xref. The options also aren't well documented there, nor are they all exposed in write_feather, so it might be best to enumerate them here.

File: r/R/dataset-format.R [hunk 9]
```diff
@@ -139,6 +139,13 @@ FileWriteOptions <- R6Class("FileWriteOptions", inherit = ArrowObject,
         dataset___ParquetFileWriteOptions__update(self,
             ParquetWriterProperties$create(...),
             ParquetArrowWriterProperties$create(...))
+      } else if (self$type == "ipc") {
+        args <- list(...)
+        dataset___IpcFileWriteOptions__update(self,
+            get_ipc_use_legacy_format(args$use_legacy_format),
+            args$codec,
```
Comment (nealrichardson) [inline r/R/dataset-format.R:146]: I think a better approach for `codec` would be to allow you to specify the parameters like `write_feather` (though fine if you want to provide your own Codec I guess). so something like this:

```suggestion
            args$codec %||% Codec$create(
              args$compression %||% "uncompressed", 
              args$compression_level %||% NA
            ),
```

Perhaps this lets you get rid of your funky `nil_value` helper too? With this change, you're always providing a Codec to C++, never `NULL`.

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: r/configure (hunk 2 above)
Comment (bkietz) [reply r/configure:204]: will do

File: r/src/arrow_cpp11.h (hunk 3 above)
Comment (bkietz) [reply r/src/arrow_cpp11.h:160]: It's a convenience for passing optional parameters; instead of writing multiple overloads of a method with different permutations of optional parameters specified those parameters can be set to `NULL` in R. It's not necessary; should I revert it?

File: r/src/arrow_metadata.h (hunk 4 above)
Comment (bkietz) [reply r/src/arrow_metadata.h:23]: 1. I can replace other instances of KeyValueMetadata wrangling to use these helpers, or I could just leave things inlined as in other files under `arrow/r/`
2. How about `KeyValueMetadata__from_strings`, `KeyValueMetadata__to_strings`?

File: r/src/recordbatchreader.cpp (hunk 5 above)
Comment (bkietz) [reply r/src/recordbatchreader.cpp:80]: I'll revert this, then

File: r/tests/testthat/test-dataset.R (hunk 7 above)
Comment (bkietz) [reply r/tests/testthat/test-dataset.R:981]: My intent was to just make the fact that we have a second destination directory for the no truncated timestamps case clearer by introducing a named variable. I don't think it changes the test's semantics at all

File: r/R/dataset-write.R (hunk 8 above)
Comment (bkietz) [reply r/R/dataset-write.R:48]: Will do

File: r/R/dataset-format.R (hunk 9 above)
Comment (bkietz) [reply r/R/dataset-format.R:146]: I'll add support for `compression, compression_level` but unfortunately `Codec$create("uncompressed")` doesn't work and we'd still need to fall back to either `nil_value` or overloads of `dataset___IpcFileWriteOptions__update`. It would also be my preference to have a non-nullptr codec in all cases but unfortunately the C++ library uses `nullptr` to indicate a lack of compression (there is no "uncompressed" codec), fixing of which is outside scope for this PR.

File: r/src/arrow_metadata.h (hunk 4 above)
Comment (nealrichardson) [reply r/src/arrow_metadata.h:23]: Yes, if these helpers are useful, let's use them everywhere. And yeah I like those names better, thanks.

File: r/R/dataset-format.R (hunk 9 above)
Comment (nealrichardson) [reply r/R/dataset-format.R:146]: so "uncompressed" codec is a nullptr? that means you can have `std::shared_ptr<arrow::Codec>` in the `dataset___IpcFileWriteOptions__update` signature, and then check for nullptr (and not R NULL)?

Comment (bkietz) [reply r/R/dataset-format.R:146]: Unfortunately, the `shared_ptr` utility already does the reverse mapping (c++ `shared_ptr<Codec>(nullptr)` -> R `NULL`) https://github.com/apache/arrow/blob/20cd08a8143eed7b48686ac3e696252dc5fdee88/r/R/arrow-package.R#L118-L124 so:

```r
uncompressed_type <- arrow:::compression_from_name("uncompressed")
uncompressed_codec <- arrow:::shared_ptr(Codec, arrow:::util___Codec__Create(uncompressed_type, NA))
assert_that(is.null(uncompressed_codec))
```

Comment (bkietz) [reply r/R/dataset-format.R:146]: Thus we'd either need to alter `arrow:::shared_ptr()` to support nullptrs or we could implement the reverse mapping and allow R `NULL` -> c++ `shared_ptr<T>(nullptr)`. For now, I'll revert `nil_value` and add overloads of `dataset___IpcFileWriteOptions__update`

File: r/tests/testthat/test-dataset.R (hunk 6 above)
Comment (bkietz) [reply r/tests/testthat/test-dataset.R:961]: Unfortunately this doesn't work; the footer metadata is distinct from the schema metadata (which is what this change would check)

Comment (nealrichardson) [reply r/tests/testthat/test-dataset.R:961]: What is footer metadata for? 

Would it make sense/be ok with you to defer wiring up footer metadata in R for this release? I don't understand it enough to advise on how to expose it in R in a way that is useful and not confusing.

File: r/R/dataset-format.R (hunk 9 above)
Comment (nealrichardson) [reply r/R/dataset-format.R:146]: Ok. Given #8256 in flight I don't think we want to mess with `shared_ptr()`, so do what you think is best.

File: r/tests/testthat/test-dataset.R (hunk 6 above)
Comment (bkietz) [reply r/tests/testthat/test-dataset.R:961]: alright

**Comment by nealrichardson** [comment]
I'm fixing the R and then will merge

**Review by nealrichardson — APPROVED** [review]
+1; macOS Python failure is unrelated (failing on master)
